## Упражнения по библиотеке Numpy

In [1]:
import numpy as np
import scipy

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
arr = np.array(np.arange(20))
arr = np.where((arr >= 3) & (arr <= 8), arr * (-1), arr)

**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
arr = np.where(arr == arr.max(), 0, arr)

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
arr1 = np.array([[1, 2], [3, 4]])
arr2 = np.array([[5, 2], [7, 8]])
np.dot(arr1, arr2)

array([[19, 18],
       [43, 38]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
arr3 = np.array([[5, 6, 5],
                [3, 4, 4],
                [2, 6, 6],
                [2, 3, 6],
                [5, 7, 7],
                [7, 3, 1],
                [3, 5, 4],
                [9, 7, 2]])
arr3

array([[5, 6, 5],
       [3, 4, 4],
       [2, 6, 6],
       [2, 3, 6],
       [5, 7, 7],
       [7, 3, 1],
       [3, 5, 4],
       [9, 7, 2]])

In [6]:
arr4 = np.array([[9, 7], [2, 7]])
arr4

array([[9, 7],
       [2, 7]])

In [7]:
arr3[np.isin(arr3, arr4.flatten()).all(axis=1)]

array([[9, 7, 2]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [8]:
arr5 = np.array([[1, 1, 2],
                [3, 2, 3],
                [2, 1, 2],
                [2, 2, 2],
                [1, 1, 1],
                [1, 2, 3],
                [3, 2, 2],
                [1, 1, 1],
                [1, 2, 1],
                [3, 1, 1]])
arr5

array([[1, 1, 2],
       [3, 2, 3],
       [2, 1, 2],
       [2, 2, 2],
       [1, 1, 1],
       [1, 2, 3],
       [3, 2, 2],
       [1, 1, 1],
       [1, 2, 1],
       [3, 1, 1]])

In [9]:
def all_equal(row):
    return len(set(row)) == 1

In [10]:
arr5[~np.apply_along_axis(all_equal, 1, arr5)]


array([[1, 1, 2],
       [3, 2, 3],
       [2, 1, 2],
       [1, 2, 3],
       [3, 2, 2],
       [1, 2, 1],
       [3, 1, 1]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [11]:
arr6 = np.array([[1, 3],
                [1, 3],
                [3, 2],
                [1, 1],
                [2, 3]])
arr6

array([[1, 3],
       [1, 3],
       [3, 2],
       [1, 1],
       [2, 3]])

In [12]:
np.unique(arr6, axis=0)

array([[1, 1],
       [1, 3],
       [2, 3],
       [3, 2]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [13]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

In [14]:
X = X.T if len(X) > len(X[0]) else X
prod = 1
for n, i in enumerate(X):
    prod *= i[n] if i[n] != 0 else 1
prod

3

In [15]:
np.where(X == 0, 1, X).diagonal().prod()

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [16]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

In [17]:
sorted(x) == sorted(y)

True

In [18]:
np.array_equal(x.sort(), y.sort())

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [19]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

In [20]:
maximum = 0
for i in range(1, len(x)):
    if (x[i - 1] == 0) & (x[i] > maximum):
        maximum = x[i]
print(maximum)

5


In [56]:
mask = (x == 0)
x[1:][mask[:-1]].max()

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [21]:
x = np.array([2, 2, 2, 3, 3, 3, 5])

In [22]:
arr_set = []
arr_count = []
x = list(x)
for i in x:
    if i not in arr_set:
        arr_set.append(i)
        arr_count.append(x.count(i))
print(arr_set)
print(arr_count)

[2, 3, 5]
[3, 3, 1]


In [23]:
uniq, counts = np.unique(x, return_counts=True)
uniq

array([2, 3, 5])

In [24]:
counts

array([3, 3, 1], dtype=int64)

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [25]:
X = np.array([[1, 2, 3], [2, 3, 4], [3, 4, 5], [2, 5, 1], [3, 5, 1]])
Y = np.array([[2, 1, 3], [3, 5, 2], [4, 2, 3], [1, 3, 2], [2, 1, 4]])

In [26]:
# # bigger arrays for better time consuming test
# X = np.array([[1, 2, 3], [2, 3, 4], [3, 4, 5], [2, 5, 1],
#               [3, 5, 1], [1, 3, 1], [2, 3, 1], [3, 4, 6],
#               [1, 4, 3], [2, 3, 1], [1, 4, 2], [4, 5, 4]])
# Y = np.array([[2, 1, 3], [3, 5, 2], [4, 2, 3], [1, 3, 2],
#               [2, 1, 4], [1, 3, 5], [2, 4, 6], [2, 1, 2],
#               [2, 1, 5], [1, 4, 4], [2, 1, 5], [5, 5, 5]])

In [27]:
# loop decision
result = []
for x_row in X:
    temp_r = []
    for y_row in Y:
        temp = 0
        for n in range(len(x_row)):
            temp += (x_row[n] - y_row[n])**2
        temp_r.append(temp**0.5)
    result.append(temp_r)
print(*result, sep='\n')

[1.4142135623730951, 3.7416573867739413, 3.0, 1.4142135623730951, 1.7320508075688772]
[2.23606797749979, 3.0, 2.449489742783178, 2.23606797749979, 2.0]
[3.7416573867739413, 3.1622776601683795, 3.0, 3.7416573867739413, 3.3166247903554]
[4.47213595499958, 1.4142135623730951, 4.123105625617661, 2.449489742783178, 5.0]
[4.58257569495584, 1.0, 3.7416573867739413, 3.0, 5.0990195135927845]


In [28]:
%%timeit
result = []
for x_row in X:
    temp_r = []
    for y_row in Y:
        temp = 0
        for n in range(len(x_row)):
            temp += (x_row[n] - y_row[n])**2
        temp_r.append(temp**0.5)
    result.append(temp_r)

233 µs ± 18.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [29]:
# numpy decision
X_lin = np.repeat(X, len(Y), axis=0)
Y_lin = np.vstack(tuple(Y for _ in range(len(Y))))
np.linalg.norm(X_lin - Y_lin, axis=1, keepdims=False).reshape(len(X), -1)

array([[1.41421356, 3.74165739, 3.        , 1.41421356, 1.73205081],
       [2.23606798, 3.        , 2.44948974, 2.23606798, 2.        ],
       [3.74165739, 3.16227766, 3.        , 3.74165739, 3.31662479],
       [4.47213595, 1.41421356, 4.12310563, 2.44948974, 5.        ],
       [4.58257569, 1.        , 3.74165739, 3.        , 5.09901951]])

In [30]:
%%timeit
X_lin = np.repeat(X, len(Y), axis=0)
Y_lin = np.vstack(tuple(Y for _ in range(len(Y))))
np.linalg.norm(X_lin - Y_lin, axis=1, keepdims=False).reshape(len(X), -1)

60 µs ± 3.11 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [31]:
# scipy decision
scipy.spatial.distance.cdist(X, Y)

array([[1.41421356, 3.74165739, 3.        , 1.41421356, 1.73205081],
       [2.23606798, 3.        , 2.44948974, 2.23606798, 2.        ],
       [3.74165739, 3.16227766, 3.        , 3.74165739, 3.31662479],
       [4.47213595, 1.41421356, 4.12310563, 2.44948974, 5.        ],
       [4.58257569, 1.        , 3.74165739, 3.        , 5.09901951]])

In [32]:
%%timeit
scipy.spatial.distance.cdist(X, Y)

14 µs ± 847 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [33]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [34]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [35]:
crunchie_cal = 60
average_calories = np.mean(calorie_stats) - crunchie_cal
average_calories

46.883116883116884

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [36]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [37]:
median_calories = np.median(calorie_stats)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [38]:
print(list(range(50, 0, -1)))

[50, 49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


In [39]:
for i in range(50, 0, -1):
    other_cal = np.percentile(calorie_stats, i)
    if i % 5 == 0:
        print(f'{100 - i}% of cereals has more than {other_cal} calories')
    if other_cal < crunchie_cal:
        nth_percentile = i+1
        break
nth_percentile

50% of cereals has more than 110.0 calories
55% of cereals has more than 110.0 calories
60% of cereals has more than 110.0 calories
65% of cereals has more than 100.0 calories
70% of cereals has more than 100.0 calories
75% of cereals has more than 100.0 calories
80% of cereals has more than 100.0 calories
85% of cereals has more than 90.0 calories
90% of cereals has more than 90.0 calories
95% of cereals has more than 70.0 calories


4

In [40]:
# better solution
from scipy import stats
stats.percentileofscore(calorie_stats, crunchie_cal)

3.896103896103896

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [41]:
round(np.sum(np.where(calorie_stats > 60, 1, 0)) / len(calorie_stats) * 100, 2)

96.1

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [42]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Мы узнали, что хлопья CrunchieMunchies менее калорийны нежели <b>96,1%</b> других хлопьев.</br>
При средней (медианной) калорийности на рынке в размере 110, данные хлопья едва ли не в <b>2(!) раза</b> менее калорийны.</br>
Стандартное отклонение в размере 19,35 показывает, что разброс в калорийности небольшой на рынке хлопьев.</br>

Полученные данные мы можем использовать, чтобы продемонстировать явное преимущество хлопьев CrunchieMunchies над большинством других хлопьев в разрезе меньшей калорийности, следовательно их можно продвигать как более здоровые и полезные для организма для людей следящих за своей фигурой.